lib/src/common/cache/cache.dart

```diff
@@ -327,23 +372,35 @@ class Cache<TIdentifier, TValue> extends Object with Disposable {
     return new Future.value();
   }
 
-  /// Run [callback] on [TValue] associated with [id].
+  /// Returns `true` and calls [callback] if there is a cached, unreleased
+  /// [TValue] associated with [id]; returns `false` otherwise.
   ///
-  /// If [TIdentifier] [TValue] pair found runs [callback] and returns true,
-  /// otherwise returns false. Will not run on released items and does not
-  /// perform a [get] or [getAsync]. As [get] or [getAsync] are not performed
-  /// calling [applyToItem] will not affect retention or removal of a
-  /// [TIdentifier][TValue] pair from the cache.
+  /// This does not perform a [get] or [getAsync], and as a result, will not
+  /// affect retention or removal of a [TIdentifier][TValue] pair from the cache.
+  /// If any callbacks are in fight on removal of [TIdentifier] [TValue] pair
+  /// [didRemove] will not event until callback has completed.
   ///
   /// If the [Cache] [isOrWillBeDisposed] then a [StateError] is thrown.
-  bool applyToItem(TIdentifier id, void callback(Future<TValue> value)) {
+  Future<bool> applyToItem(
+      TIdentifier id, Future<Null> callback(Future<TValue> value)) {
     _log.finest('applyToItem id: $id');
     _throwWhenDisposed('applyToItem');
     if (_isReleased[id] != false) {
-      return false;
+      return new Future.value(false);
+    }
+    final callbackFuture = callback(_cache[id]);
+    if (callbackFuture is Future<Null>) {
+      _applyToItemCallBacks.putIfAbsent(id, () => <Future<Null>>[]);
+      _applyToItemCallBacks[id].add(callbackFuture);
```

Comment:
From above: if `<Null>` is dropped or it's made dynamic, it should also be dropped here and on the typing for `_applyToItemCallBacks`.
```dart
if (callbackFuture is Future) {
  _applyToItemCallBacks.putIfAbsent(id, () => <Future>[]);
  _applyToItemCallBacks[id].add(callbackFuture);